Dataloader api

galalen opened the pull request:
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

galalen committed:
add base files

galalen committed:
base dataloader api

galalen committed:
impl dataloader and update samplers

galalen committed:
update demo and add test cases for smaplers

galalen committed:
update dataloader example

galalen committed:
fix dataloder indices update, update training example

galalen committed:
refator dataloader classes, add unit tests

galalen committed:
add more unit tests

mccorby commented on syft/src/main/java/org/openmined/syft/data/fetchers/BaseDatasetFetcher.kt:
```diff
@@ -0,0 +1,31 @@
+import org.pytorch.IValue
+import org.pytorch.Tensor
+
+open class BaseDatasetFetcher(val dataset: Dataset) : Fetcher {
```
Is this class supposed to be extended?

mccorby commented on syft/src/main/java/org/openmined/syft/data/fetchers/BaseDatasetFetcher.kt:
```diff
@@ -0,0 +1,31 @@
+
+open class BaseDatasetFetcher(val dataset: Dataset) : Fetcher {
+
+    override fun fetch(indices: List<Int>): Pair<IValue, IValue> {
```
```
    indices.forEach { index ->
        val pair = dataset.getItem(index)
        data.add(pair.first)
        labels.add(pair.second)
    }

   val batchSize = indices.size.toLong()
    val xFeatureLength = data[0].first.shape.last()
    val yFeatureLength = labels[0].second.shape.last()

return Pair(
            IValue.from(Tensor.fromBlob(data.flatten().toFloatArray(), longArrayOf(batchSize, xFeatureLength))),
            IValue.from(Tensor.fromBlob(labels.flatten().toFloatArray(), longArrayOf(batchSize, yFeatureLength)))
        )
```

galalen commented on syft/src/main/java/org/openmined/syft/data/DataLoader.kt:
```diff
@@ -0,0 +1,68 @@
+    override fun iterator(): Iterator<Pair<IValue, IValue>> = BaseDataLoaderIterator(this)
+}
+
+open class BaseDataLoaderIterator(dataLoader: DataLoader) : Iterator<Pair<IValue, IValue>> {
```
In case there is another implementation to the iterator, but **YAGNI**,
I'll modify it.

galalen commented on syft/src/main/java/org/openmined/syft/data/KTensor.kt:
```diff
@@ -1,11 +1,19 @@
+ * @property flattenedArray the actual data
+ * @property shape data shape
+ * */
+data class KTensor(val flattenedArray: FloatArray, val shape: LongArray) {
```
SyftTensor is perfect if we would not need the name for something else.
But SyftTensor is the right choice.

vkkhare replied:
What functionality are we adding over a usual tensor for needing another class? as far as I can see, we have shape and data array. So SyftTensor at this moment is a serialization tensor class so it can be renamed to SyftProtoTensor.

galalen commented on syft/src/main/java/org/openmined/syft/data/Dataset.kt:
```diff
@@ -0,0 +1,23 @@
+    /**
+     * This method is called to fetch a data sample for a given key.
+     */
+    fun getItem(index: Int) : Pair<KTensor, KTensor>
```
KTensor (SyftTensor) is just a util class to help create IValue later, and keep the data and the shape separated

mccorby replied:
Let's leave it like this by the time being.
Could you please open an issue to give it a thought to the naming of these objects?

t-vi replied:
KTensor (SyftTensor) is just a util class to help create IValue later, and keep the data and the shape separated

I accidentaly saw the namming issue on slack. Out of curiosity, from the PR, what is missing from it that keeps you from just using Tensor here? This would copy the data earlier but from the uses of KTensor in this PR it would be covered by Tensor (with the shape being in the tensor but available through .shape() AFAIU)
Maybe it would not be too hard to provide the necessary bindings.

galalen replied:
Make sense, I thought constructing the Tensor would be hard for the user at first, but it's not, and it's better not to introduce another data type here

galalen committed:
merge coroutines api

vkkhare commented on syft/src/main/java/org/openmined/syft/data/DataLoader.kt:
```diff
@@ -0,0 +1,68 @@
+ */
+class DataLoader(var dataset: Dataset,
+                 var batchSize: Int = 1,
+                 var shuffle: Boolean = false,
```
Pytorch does it like this. I think it would be good to go this way.

mccorby replied:
We should not follow PyTorch to this extreme. IT is our API!
But let's keep it by now.

vkkhare replied:
Ahh I meant having a sampler injected is how pytorch does it

mccorby replied:
We should follow PyTorch to this extreme 😬

galalen commented on syft/src/main/java/org/openmined/syft/data/DataLoader.kt:
```diff
@@ -0,0 +1,68 @@
+        dropLast
+    )
+
+    fun indexSampler() = batchSampler
```
same as pytorch api

vkkhare replied:
Instead of making indexSampler a method, make it an overridable property of the base class then you can just do `override val indexSampler = BatchSampler(sampler,batchSize,dropLast)`

galalen committed:
merge dataloader api with coroutines, replace KTensor with pytorch tensor

galalen commented on syft/src/main/java/org/openmined/syft/execution/SyftJob.kt:
```diff
@@ -141,51 +142,53 @@ class SyftJob internal constructor(
 
                 emit(TrainingState.Epoch(step + 1))
-                val batchData = syftDataLoader.loadDataBatch(batchSize)
+                val dataLoader = DataLoader(dataset, batchSize)
```
Nope, I'm gonna take out, it will be initialize once, then then I'll feed the data loader to train method.

galalen committed:
add reset to dataloader, make sampler injectable in dataloader

galalen committed:
Unique data source per job (#278)

* update job local datasource with helper function

add unit test for helper functions on job local data source

refine job repo

* refactor JobDataModel to JobId

* sperate job data sources

* sperate job data sources

update annotation

* add factory method to job repository

update job

update idea

udpate idea

mccorby reviewed (approved):
LGTM
Nice job!

vkkhare commented on syft/src/main/java/org/openmined/syft/data/samplers/RandomSampler.kt:
```diff
@@ -0,0 +1,19 @@
+class RandomSampler(private val dataset: Dataset) :
+    Sampler {
+
+    override fun indices() = (0 until dataset.length()).shuffled(Random()).toList()
```
List(dataset.length()) { it }

vkkhare commented on syft/src/main/java/org/openmined/syft/data/DataLoader.kt:
```diff
@@ -0,0 +1,67 @@
+
+class DataLoaderIterator(dataLoader: DataLoader) : Iterator<Pair<IValue, IValue>> {
+
+    private val indexSampler = dataLoader.indexSampler()
```
again use it as `dataloader.indexSampler` as property access

vkkhare commented on syft/src/main/java/org/openmined/syft/data/Dataset.kt:
```diff
@@ -0,0 +1,25 @@
+    /**
+     * This method is called to fetch a data sample for a given key.
+     */
+    fun getItem(index: Int) : Pair<Tensor, Tensor>
```
instead of a Pair make it a list/dict of Tensor/IValue

vkkhare commented on syft/src/main/java/org/openmined/syft/data/fetchers/BaseDatasetFetcher.kt:
```diff
@@ -0,0 +1,30 @@
+import org.pytorch.IValue
+import org.pytorch.Tensor
+
+open class BaseDatasetFetcher(val dataset: Dataset) : Fetcher {
```
From the looks of it this class is only responsible for collating the independent data points into batch. So why do we need a separate class for this? Can't there be a collating function in the dataloader class? I agree this collating method should be overrideable and the user can simply pass this method as argument if he wants

vkkhare commented on syft/src/main/java/org/openmined/syft/data/fetchers/Fetcher.kt:
```diff
@@ -0,0 +1,9 @@
+
+import org.pytorch.IValue
+
+interface Fetcher {
```
seems redundant class, see explanation above.

vkkhare commented on syft/src/main/java/org/openmined/syft/data/samplers/BatchSampler.kt:
```diff
@@ -0,0 +1,45 @@
+    private val indices = sampler.indices()
+    private var currentIndex = 0
+
+    override fun indices() = when {
```
again indices should be a property rather than a method

vkkhare commented on syft/src/main/java/org/openmined/syft/data/samplers/Sampler.kt:
```diff
@@ -0,0 +1,14 @@
+ */
+interface Sampler {
+
+    fun indices(): List<Int>
```
make these properties not methods

github-actions[bot] committed:
PR for auto doc update (#298)

Co-authored-by: vkkhare <vkkhare@users.noreply.github.com>

galalen committed:
update dataset return type

galalen committed:
update dataloader

vkkhare commented on syft/src/main/java/org/openmined/syft/execution/SyftJob.kt:
```diff
@@ -37,16 +37,14 @@ private const val TAG = "SyftJob"
 class SyftJob internal constructor(
-    modelName: String,
-    version: String? = null,
+    val modelName: String,
```
Why val here?

vkkhare commented on syft/src/main/java/org/openmined/syft/execution/SyftJob.kt:
```diff
@@ -70,21 +68,22 @@ class SyftJob internal constructor(
+    internal val jobRepository = JobRepository.create(config, modelName, version)
+
     internal var cycleStatus = AtomicReference(CycleStatus.APPLY)
+
```
Unnecessary spaces

vkkhare commented on syft/src/main/java/org/openmined/syft/execution/SyftJob.kt:
```diff
@@ -293,12 +290,12 @@ class SyftJob internal constructor(
         )
         val oldState =
-                SyftState.loadSyftState("${config.filesDir}/models/${model.pyGridModelId}.pb")
+                SyftState.loadSyftState("${jobRepository.getModelsPath()}/${model.pyGridModelId}.pb")
```
I think modelsPath already has model id appended

vkkhare commented on syft/src/main/java/org/openmined/syft/data/samplers/BatchSampler.kt:
```diff
@@ -16,30 +16,32 @@ class BatchSampler(
 ) : Sampler {
 
-    private val indices = sampler.indices()
+    private val mIndices = sampler.indices
```
Rename sampler to something else

vkkhare replied:
check this

vkkhare commented on syft/src/main/java/org/openmined/syft/data/Dataset.kt:
```diff
@@ -20,6 +20,6 @@ interface Dataset {
      * This method is called to fetch a data sample for a given key.
      */
-    fun getItem(index: Int) : Pair<Tensor, Tensor>
+     fun getItem(index: Int) : List<IValue>
```
Remove space

vkkhare commented on syft/src/main/java/org/openmined/syft/data/DataLoader.kt:
```diff
@@ -23,43 +22,62 @@ class DataLoader(var dataset: Dataset,
+        return fetch(indices)
+    }
+
+    private fun fetch(indices: List<Int>): List<IValue> {
```
Fetch function should be a part of dataloader and should be overridable by the user

vkkhare commented on syft/src/main/java/org/openmined/syft/data/DataLoader.kt:
```diff
@@ -23,43 +22,62 @@ class DataLoader(var dataset: Dataset,
 
-    override fun iterator(): Iterator<Pair<IValue, IValue>> = iterator
+    override fun iterator(): Iterator<List<IValue>> = iterator
 
```
Property not method

vkkhare commented on docs/syft/org.openmined.syft/index.md:
```diff
@@ -6,4 +6,4 @@
 | Name | Summary |
 |---|---|
-| [Syft](-syft/index.md) | This is the main syft worker handling creation and deletion of jobs. This class is also responsible for monitoring device resources via DeviceMonitor`class Syft : Disposable` |
+| [Syft](-syft/index.md) | This is the main syft worker handling creation and deletion of jobs. This class is also responsible for monitoring device resources via DeviceMonitor`class Syft` |
```
Remove all these docs

vkkhare commented on syft/src/main/java/org/openmined/syft/execution/JobId.kt:
```diff
@@ -0,0 +1,31 @@
+ * @property version The model version in PyGrid
+ */
+data class JobId(
+    val modelName: String,
```
Maybe make a jobModel class that stores jobId, modelname, model version, plans and protocols.

vkkhare commented on syft/src/main/java/org/openmined/syft/execution/SyftJob.kt:
```diff
@@ -69,21 +68,22 @@ class SyftJob internal constructor(
-                    JobRemoteDataSource(config.getDownloader())
-                )
+                config
             )
```
Job repository should be injected and so is job model

vkkhare commented on syft/src/main/java/org/openmined/syft/execution/SyftJob.kt:
```diff
@@ -69,21 +68,22 @@ class SyftJob internal constructor(
     private val isDisposed = AtomicBoolean(false)
 
     private val plans = ConcurrentHashMap<String, Plan>()
+
```
Send all of these variables into jobModel. It should store plans, protocols, dispose state and requires speed test along with model name, model version and jobId

galalen committed:
compose job specific info into jobModel, add AbstractDataLoader

galalen committed:
merge with dev

vkkhare commented on .idea/codeStyles/Project.xml:
```diff
@@ -9,21 +9,11 @@
-        <value>
-          <package name="" alias="false" withSubpackages="true" />
-          <package name="java" alias="false" withSubpackages="true" />
-          <package name="javax" alias="false" withSubpackages="true" />
```
remove these

vkkhare commented on docs/syft/index.md:
```diff
@@ -1,16 +1,9 @@
 |---|---|
-| [org.openmined.syft](org.openmined.syft/index.md) |  |
-| [org.openmined.syft.domain](org.openmined.syft.domain/index.md) |  |
-| [org.openmined.syft.execution](org.openmined.syft.execution/index.md) |  |
```
remove all the docs

vkkhare commented on syft/src/main/java/org/openmined/syft/data/Dataset.kt:
```diff
@@ -0,0 +1,25 @@
+    /**
+     * This method is called to return the size of the dataset, needs to be overridden.
+     */
+    fun length() : Int
```
should be a property

vkkhare reviewed (commented):
Some cosmetic changes and it is good to go!

galalen committed:
dataloader api

galalen committed:
add tests to data loader iterator

add docs

vkkhare reviewed (commented):
You didn't need to delete these files rather just reset them to whatever they were without your changes. Now the PR shows 294 files changed ;p
